util/gpool/spmc/spmspool.go

```diff
@@ -0,0 +1,362 @@
+			p.cond.Broadcast()
+		}
+	}
+}
+
+// Tune changes the capacity of this pool, note that it is noneffective to the infinite or pre-allocation pool.
+func (p *Pool[T, U, C]) Tune(size int) {
+	capacity := p.Cap()
+	if capacity == -1 || size <= 0 || size == capacity || p.options.PreAlloc {
+		return
+	}
+	p.capacity.Store(int32(size))
+	if size > capacity {
+		if size-capacity == 1 {
+			p.cond.Signal()
+			return
+		}
+		p.cond.Broadcast()
+	}
+}
+
+// Running returns the number of workers currently running.
+func (p *Pool[T, U, C]) Running() int {
+	return int(p.running.Load())
+}
+
+// Free returns the number of available goroutines to work, -1 indicates this pool is unlimited.
+func (p *Pool[T, U, C]) Free() int {
+	c := p.Cap()
+	if c < 0 {
+		return -1
+	}
+	return c - p.Running()
+}
+
+// Waiting returns the number of tasks which are waiting be executed.
+func (p *Pool[T, U, C]) Waiting() int {
+	return int(p.waiting.Load())
+}
+
+// IsClosed indicates whether the pool is closed.
+func (p *Pool[T, U, C]) IsClosed() bool {
+	return p.state.Load() == gpool.CLOSED
+}
+
+// Cap returns the capacity of this pool.
+func (p *Pool[T, U, C]) Cap() int {
+	return int(p.capacity.Load())
+}
+
+func (p *Pool[T, U, C]) addRunning(delta int) {
+	p.running.Add(int32(delta))
+}
+
+func (p *Pool[T, U, C]) addWaiting(delta int) {
+	p.waiting.Add(int32(delta))
+}
+
+// Release closes this pool and releases the worker queue.
+func (p *Pool[T, U, C]) Release() {
+	if !p.state.CompareAndSwap(gpool.OPENED, gpool.CLOSED) {
+		return
+	}
+	p.lock.Lock()
+	p.workers.reset()
+	p.lock.Unlock()
+	// There might be some callers waiting in retrieveWorker(), so we need to wake them up to prevent
+	// those callers blocking infinitely.
+	p.cond.Broadcast()
+}
+
+func isClose(exitCh chan struct{}) bool {
+	select {
+	case <-exitCh:
+		return true
+	default:
+	}
+	return false
+}
+
+// ReleaseAndWait is like Release, it waits all workers to exit.
+func (p *Pool[T, U, C]) ReleaseAndWait() {
+	if p.IsClosed() || isClose(p.stopCh) {
+		return
+	}
+
+	close(p.stopCh)
+	p.Release()
+	for {
+		if p.Running() == 0 && p.heartbeatDone.Load() == 1 {
+			return
+		}
+	}
+}
+
+// SetConsumerFunc is to set ConsumerFunc which is to process the task.
+func (p *Pool[T, U, C]) SetConsumerFunc(consumerFunc func(T, C) U) {
+	p.consumerFunc = consumerFunc
+}
+
+// AddProduce is to add Produce.
+func (p *Pool[T, U, C]) AddProduce(task T, constArg C) (<-chan U, TaskController) {
+	result := make(chan U, 1)
+	var wg sync.WaitGroup
+	tc := TaskController{
+		wg: &wg,
+	}
+	taskCh := make(chan T)
+	taskBox := taskBox[T, U, C]{
+		task:      taskCh,
+		constArgs: &constArg,
+		wg:        &wg,
+		resultCh:  result,
+	}
+	p.taskCh <- &taskBox
+	taskCh <- task
+	close(taskCh)
+	return result, tc
+}
+
+// AddProducer is to add producer.
+func (p *Pool[T, U, C]) AddProducer(producer func() (T, error), constArg C, size int) (<-chan U, TaskController) {
```

Comment:
Could we add the arg of `producer` with `[]T`? Then `producer` can add a batch of `T`